chunk_store.go

```diff
@@ -265,39 +265,107 @@ func (c *AWSChunkStore) Put(ctx context.Context, chunks []wire.Chunk) error {
 		return err
 	}
 
-	// TODO: parallelise
+	err = c.putChunks(userID, chunks)
+	if err != nil {
+		return err
+	}
+
+	return c.updateIndex(userID, chunks)
+}
+
+// putChunks writes a collection of chunks to S3 in parallel.
+func (c *AWSChunkStore) putChunks(userID string, chunks []wire.Chunk) error {
+	incomingErrors := make(chan error)
 	for _, chunk := range chunks {
-		err := timeRequest("Put", s3RequestDuration, func() error {
-			var err error
-			_, err = c.s3.PutObject(&s3.PutObjectInput{
-				Body:   bytes.NewReader(chunk.Data),
-				Bucket: aws.String(c.bucketName),
-				Key:    aws.String(chunkName(userID, chunk.ID)),
-			})
-			return err
-		})
+		go func(chunk wire.Chunk) {
+			incomingErrors <- c.putChunk(userID, &chunk)
+		}(chunk)
+	}
+
+	var lastErr error
+	for range chunks {
+		err := <-incomingErrors
 		if err != nil {
-			return err
+			lastErr = err
+		}
+	}
+	return lastErr
+}
+
+// putChunk puts a chunk into S3.
+func (c *AWSChunkStore) putChunk(userID string, chunk *wire.Chunk) error {
+	err := timeRequest("Put", s3RequestDuration, func() error {
+		var err error
+		_, err = c.s3.PutObject(&s3.PutObjectInput{
+			Body:   bytes.NewReader(chunk.Data),
+			Bucket: aws.String(c.bucketName),
+			Key:    aws.String(chunkName(userID, chunk.ID)),
+		})
+		return err
+	})
+	if err != nil {
+		return err
+	}
+
+	if c.chunkCache != nil {
+		if err = c.chunkCache.StoreChunkData(userID, chunk); err != nil {
+			log.Warnf("Could not store %v in chunk cache: %v", chunk.ID, err)
 		}
+	}
+	return nil
+}
 
-		if c.chunkCache != nil {
-			if err = c.chunkCache.StoreChunkData(userID, &chunk); err != nil {
-				log.Warnf("Could not store %v in chunk cache: %v", chunk.ID, err)
-			}
+func (c *AWSChunkStore) updateIndex(userID string, chunks []wire.Chunk) error {
+	writeReqs, err := c.calculateDynamoWrites(userID, chunks)
+	if err != nil {
+		return err
+	}
+
+	// Batch the requests.
+	batches := [][]*dynamodb.WriteRequest{}
+	for i := 0; i < len(writeReqs); i += dynamoMaxBatchSize {
+		var reqs []*dynamodb.WriteRequest
+		if i+dynamoMaxBatchSize > len(writeReqs) {
+			reqs = writeReqs[i:]
+		} else {
+			reqs = writeReqs[i : i+dynamoMaxBatchSize]
+		}
+		batches = append(batches, reqs)
+	}
```

Comment:
I've moved the batching logic to its own method, and left the parallel logic in `updateIndex`.

Reason I did it this way is that the batching logic at its heart is a simple generic function[1], so putting it by itself makes it easy to test (optimistically), and at the call site, makes it obvious that it can't fail.


[1] e.g.

```python
def batch(xs, n):
    i = 0
    num_items = len(xs)
    while i < xs:
        yield xs[i:i+n]
        i += n
```